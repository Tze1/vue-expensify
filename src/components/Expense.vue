<template>
  <div class="expense component">
    <span class="expense-createdat">{{ createdAtDisplay }}</span>
    <span class="expense-description">{{ expense.description }}</span>
    <span class="expense-amount">{{ amountDisplay }}</span>
    <router-link :to="editExpenseLink">edit</router-link>
  </div>
</template>

<script>
  import moment from 'moment';
  import numeral from 'numeral';

  export default {
    name: 'Expense',
    props: {
      expense: {
        type: Object,
        required: true,
      },
    },
    computed: {
      editExpenseLink: function () {
        return '/edit/:' + this.expense.id;
      },
      createdAtDisplay: function () {
        return moment(this.expense.createdAt).format('MM/DD/YYYY');
      },
      amountDisplay: function () {
        return numeral(this.expense.amount / 100).format('$0,0.00');
      },
    },
  };
</script>

<style scope lang="scss">
  @import '../styles/app.scss';

  .expense.component {
    border-bottom: 1px dotted $gray4;
    display: flex;
    flex-flow: row nowrap;
    font-size: 1.6rem;
    justify-content: flex-start;
    margin-bottom: $space-sm / 2;
    padding-bottom: $space-sm / 2;

    &:nth-child(odd) {
      background-color: rgba(0, 0, 0, .65);
    }

    &:first-child {
      border-top: 1px solid $gray4;
      padding-top: $space-sm / 2;
      margin-top: $space-sm / 2;
    }

    &:last-child {
      border-bottom: 1px solid $gray4;
    }

    > * {
      flex: 0 0 auto;
      padding-right: .8rem;

      &.expense-description {
        flex-grow: 1;
      }

      &:last-child {
        padding-right: 0;
      }
    }
  }
</style>
