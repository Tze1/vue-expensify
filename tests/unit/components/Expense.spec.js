import { shallowMount, mount, createLocalVue, RouteLinkStub } from '@vue/test-utils';
import Vue from 'vue';
import Vuex from 'vuex';
import testExpenses from '../../fixtures/testExpenses';
import Expense from '@/components/Expense';

/* Even though vue-test-utils docs says
* Vue.nextTick is not needed to wait for DOM updates,
* it's needed below for EditExpense method-calls that in turn call
* bootstrap-vue-components' show/hide methods.
*/

const localVue = createLocalVue();
localVue.use(Vuex);

describe('Expense.vue', () => {
  const dispatchStub = jest.fn();
  const pushStub = jest.fn();
  const $store = {
    state: {
      expenses: testExpenses,
    },
    dispatch: dispatchStub,
  };
  const $router = {
    push: pushStub,
  };
  const $route = {
    name: 'Expense',
    params: { id: '1' },
  };
  const testExpense = testExpenses[0];
  const testEditLinkPath = '/edit/' + testExpenses[0].id;
  const wrapper = shallowMount(Expense, {
    stubs: [ 'router-link' ],
    propsData: {
      expense: testExpense,
    },
  });

  it('renders Expense component properly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders editExpenseLink path properly', () => {
    expect(wrapper.find('router-link-stub').attributes('to')).
      toBe(testEditLinkPath);
  });

  it('displays confirm-modal on remove-button click', () => {
    const wrapper = mount(Expense, {
    stubs: [ 'router-link' ],
      mocks: {
        $store,
        $router,
        $route,
      },
      propsData: {
        expense: testExpense,
      },
    });

    wrapper.vm.$refs.confirmRemoveModal.hide();

    Vue.nextTick(() => {
      wrapper.find('.expense-remove').trigger('click');
    });

    Vue.nextTick(() => {
      // modal should no longer have style="display:none;" once displayed.
      expect(wrapper.find('.expense-confirm .modal').
        attributes('style')).toMatch('');
      wrapper.vm.$refs.confirmRemoveModal.hide();
    });

  });

  it('hides confirm-modal on modal-cancel-button click', () => {
    const wrapper = mount(Expense, {
    stubs: [ 'router-link' ],
      mocks: {
        $store,
        $router,
        $route,
      },
      propsData: {
        expense: testExpense,
      },
    });

    wrapper.vm.$refs.confirmRemoveModal.show();
    // modal should no longer have style="display:none;" once displayed.

    Vue.nextTick(() => {
      wrapper.find('.modal-footer .btn-secondary').trigger('click');
      // modal should now have style="display:none;" once displayed.
    });

    Vue.nextTick(() => {
      expect(wrapper.find('.expense-confirm .modal').
        attributes('style')).toMatch('display: none');
    });
  });

  it('hides confirm-modal & dispatches removeExpense action properly on modal-remove-button click', () => {
    const id = $route.params.id;
    const wrapper = mount(Expense, {
      localVue,
    stubs: [ 'router-link' ],
      mocks: {
        $store,
        $router,
        $route
      },
      propsData: {
        expense: testExpense,
      },
    });

    wrapper.vm.onRemove();  // displays confirm-modal.

    Vue.nextTick(() => {
      wrapper.find('.modal-footer .btn-danger').trigger('click');
      // confirm-modal should now be hidden.
    });

    Vue.nextTick(() => {
      expect(wrapper.find('.expense-confirm .modal').
        attributes('style')).toMatch('display: none');
      expect(dispatchStub).toHaveBeenCalledWith('removeExpense', id);
    });
  });
});
