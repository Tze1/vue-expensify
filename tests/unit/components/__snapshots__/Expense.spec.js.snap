// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expense.vue renders Expense component properly 1`] = `
<div class="expense component">
  <div class="expense-wrapper"><span class="expense-createdat">12/31/1969</span> <span class="expense-description">Gum</span> <span class="expense-amount">$1.95</span>
    <router-link-stub to="/edit/1" title="Edit Expense" aria-label="Edit Expense" class="expense-edit">
      <font-awesome-icon-stub icon="edit"></font-awesome-icon-stub>
    </router-link-stub> <button title="Remove Expense" aria-label="Remove Expense" class="expense-remove">
      <font-awesome-icon-stub icon="trash-alt"></font-awesome-icon-stub>
    </button></div>
  <b-modal-stub title="Confirm Remove" titletag="h5" size="md" centered="true" buttonsize="" nocloseonbackdrop="true" hideheaderclose="true" headercloselabel="Close" canceltitle="Cancel" oktitle="Yes, remove" cancelvariant="secondary" okvariant="danger" class="expense-confirm">
    <p class="confirm-message">Are you sure? Removing cannot be undone.</p>
  </b-modal-stub>
</div>
`;
